### PR TITLE
support for static route yang model 

### PR DESCRIPTION
**Pull request by UmaMaven**
Signed-off-by: Uma Ramanathan uma.ramanathan@mavenir.com
#### Why I did it
Support yang model for static route
#### How I did it
Added a yang file for static route configuration 
#### How to verify it
Add static routes with gateway/interface/blackhole as nexthop with VRF capability

#11498

### CODE REVIEW

**Comment by ganglyu** [comment]
Please update setup.py and sample_config_db.json, for example, https://github.com/Azure/sonic-buildimage/pull/10642

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            "Virtual Routing Instance name as key";
+        }
+        leaf nexthop {
+          type inet:ip-address;
+          description
+            "The next-hop that is to be used for the
+             static route as IP address. When interface needs to be
+             specified, use 0.0.0.0 as leaf value";
+        }
+        leaf ifname {
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:58]: why not use leafref?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            }
+            type leafref {
+              path "/vrf:sonic-vrf/vrf:VRF/vrf:VRF_LIST/vrf:name";
+            }
+          }
+          description
+            "VRF name of the nexthop. This is for vrf leaking";
+        }
+        leaf blackhole {
+          type string;
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:92]: blackhole can be true or false.
why not use boolean?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            pattern 'Ethernet([1-3][0-9]{3}|[1-9][0-9]{2}|[1-9][0-9]|[0-9])' {
+              error-message "Invalid interface name";
+              error-app-tag "interface-name-invalid";
+            }
+          }
+          description
+            "When interface is specified, forwarding happens through it";
+        }
+        leaf distance {
+          type string;
```
Comment (ganglyu) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:69]: why not use int?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  container sonic-static-route {
+    container STATIC_ROUTE {
+      description
+        "STATIC_ROUTE part of config_db.json";
+      list STATIC_ROUTE_LIST {
+        key "vrf_name prefix";
+        leaf prefix {
+          type stypes:sonic-ip4-prefix;
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:35]: We should support ipv6 prefix as well.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+          description
+            "prefix is the destination IP address, as key";
+        }
+        leaf vrf_name {
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:39]: Move this as first field to match key order.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            }
+            type leafref {
+              path "/vrf:sonic-vrf/vrf:VRF/vrf:VRF_LIST/vrf:name";
+            }
+          }
+          description
+            "Virtual Routing Instance name as key";
+        }
+        leaf nexthop {
+          type inet:ip-address;
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:52]: ECMP not supported? we need comma seperated NH list.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 1 above)
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:58]: We need comma separated interface names to support ECMP paths, so, we need custom valdiation and leafref wont help.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 3 above)
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:69]: The backend expects the distance as string, please add string pattern to allow only the valid range.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+            "Administrative Distance (preference) of the entry.  The
+             preference defines the order of selection when multiple
+             sources (protocols, static, etc.) contribute to the same
+             prefix entry.  The lower the preference, the more preferable the
+             prefix is.  When this value is not specified, the preference is
+             inherited from the default preference of the implementation for
+             static routes.";
+        }
+        leaf nexthop-vrf {
+          type union {
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:80]: This one again needs to be comma separated VRF names to support ECMP case.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 2 above)
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:92]: Backend expect string, please add pattern to allow true/false strings.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 4 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:35]: Can add it here in yang, however would not be able to test it as a solution.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 6 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:52]: Can make the yang file support it. However we would not be able to test it on our solution.

Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:52]: frrcfgd already supports ECMP support for static route, please check the ECMP route manually in FRR.

**Review by ganglyu — APPROVED** [review]

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+          }
+          description
+            "The next-hop that is to be used for the
+             static route as IP address. When interface needs to be
+             specified, use 0.0.0.0 as leaf value";
+        }
+        leaf ifname {
+          type string {
+	   pattern "(Ethernet([1-3][0-9]{3}|[1-9][0-9]{2}|[1-9][0-9]|[0-9])+,)*(Ethernet([1-3][0-9]{3}|[1-9][0-9]{2}|[1-9][0-9]|[0-9])+)";
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:62]: We need the pattern matching for Ethernet, Portchannel, sub-interfaces and VLAN..etc..please check all possible interfaces and do the pattern matching OR leave as string for now and add a custom validation later.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+             preference defines the order of selection when multiple
+             sources (protocols, static, etc.) contribute to the same
+             prefix entry.  The lower the preference, the more preferable the
+             prefix is.  When this value is not specified, the preference is
+             inherited from the default preference of the implementation for
+             static routes.";
+        }
+        leaf nexthop-vrf {
+          type string {
+            pattern "(((Vrf[a-zA-Z0-9_-]+)|(default)),)*((Vrf[a-zA-Z0-9_-]+)|(default))";
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:83]: How about mgmt VRF?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
...1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])),)*)(((([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])))))$';
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:53]: I'm not sure this covers all combinations of IP addresses, we can better have customer validations.

Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:53]: What combinations of IP addresses? 0.0.0.0 is covered. and I tested for couple of IP addresses. Anything else you have in mind?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 9 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:83]: You mean, "mgmt" as a string?

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 8 above)
Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:62]: Okay. I will leave it as String & add custom validation later.

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 9 above)
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:83]: Yes

File: src/sonic-yang-models/yang-models/sonic-static-route.yang (hunk 10 above)
Comment (venkatmahalingam) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:53]: Will this regexp work for ipv4 embeded ipv6 addresses e.g " ::ffff:192.168.0.1" ? I'm ok if you have covered all combinations in the test cases.

**Review by prsunny — APPROVED** [review]

**Comment by venkatmahalingam** [comment]
@UmaMaven If you could push the changes for the last few comments, we can merge this PR.

**Comment by UmaMaven** [comment]
> @UmaMaven If you could push the changes for the last few comments, we can merge this PR.

Hi Venkat, I have uploaded the files with the last few comments as well. 
Thanks

File: src/sonic-yang-models/yang-models/sonic-static-route.yang [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    container STATIC_ROUTE {
+      description
+        "STATIC_ROUTE part of config_db.json";
+      list STATIC_ROUTE_LIST {
+        key "vrf_name prefix";
+        leaf vrf_name {
+          type union {
+            type string {
+              pattern 'default';
+            }
```
Comment (venkatmahalingam) [inline src/sonic-yang-models/yang-models/sonic-static-route.yang:38]: Please add 'mgmt' string here as well.

Comment (UmaMaven) [reply src/sonic-yang-models/yang-models/sonic-static-route.yang:38]: @venkatmahalingam Please approve. Thanks

**Review by venkatmahalingam — APPROVED** [review]

**Review by prsunny — APPROVED** [review]

**Review by venkatmahalingam — APPROVED** [review]

**Pull request merged**